src/Components/Components/src/PersistentState/ComponentStatePersistenceManager.cs

```diff
@@ -72,24 +94,53 @@ async Task PauseAndPersistState()
                 // the next store can start with a clean slate.
                 foreach (var store in compositeStore)
                 {
-                    await PersistState(store);
+                    var result = await TryPersistState(store);
+                    if (!result)
+                    {
+                        break;
+                    }
                     _currentState.Clear();
                 }
             }
             else
             {
-                await PersistState(store);
+                await TryPersistState(store);
             }
 
             State.PersistingState = false;
             _stateIsPersisted = true;
         }
 
-        async Task PersistState(IPersistentComponentStateStore store)
+        async Task<bool> TryPersistState(IPersistentComponentStateStore store)
         {
-            await PauseAsync(store);
+            if (!await TryPauseAsync(store))
+            {
+                _currentState.Clear();
+                return false;
+            }
+
             await store.PersistStateAsync(_currentState);
```

Comment:
`PersistStateAsync` can throw. Then, instead of returning false we prefer to propagate `InvalidOperationException("State already persisted.")`, even though the method has `Try` in the name?

Reply:
Fine, `PersistStateAsync` from this file does not come from `IPersistentComponentStateStore` implementation, the only  `PersistStateAsync` in a `IPersistentComponentStateStore`-based class I could find is
https://github.com/dotnet/aspnetcore/blob/12d57ddde6bd2f47757364252596cfa89df2ef60/src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.PrerenderingState.cs#L269
and there we don't throw for real. The commend can be ignored.

Reply:
Yep, we are just propagating when a user callback fails because we don't want to persist any state moving forward in that scenario. If the store throws that's fine, and we want that exception to bubble up and propagate.